### DISCUSSION TITLE
Is it possible to use import std; with CMake?

### DISCUSSION THREAD

**Discussion by Jaakkko** [android/ndk#2111] [Q&A] [answered]
I'm using
- Android NDK 28.0.12674087
- CMake 3.31.2

According to [cmake-cxxmodules](https://cmake.org/cmake/help/latest/manual/cmake-cxxmodules.7.html) `import std` can be used if Clang 18.1.2 or newer with `-stdlib=libc++` is used. And if I'm understanding the [NDK](https://developer.android.com/ndk/guides/cmake#android_stl) documentation correctly `libc++` should be enabled by default. I'm meeting all these requirements (Clang version is 19.0.0).

However, I'm getting this error:
```
CMake Error in CMakeLists.txt:
  The "CXX_MODULE_STD" property on the target "myapplication" requires that
   the "__CMAKE::CXX23" target exist, but it was not provided by the
   toolchain.  Reason:
     Only `libc++` is supported
```

My `CMakeLists.txt` looks like this:
```CMake
cmake_minimum_required(VERSION 3.31.2 FATAL_ERROR)

set(CMAKE_EXPERIMENTAL_CXX_IMPORT_STD "0e5b6991-d74f-4b3d-a41c-cf096e0b2508")

project("myapplication" LANGUAGES CXX)

set(CMAKE_CXX_MODULE_STD ON)
set(CMAKE_CXX_STANDARD 23)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS ON)

add_library(${CMAKE_PROJECT_NAME} SHARED myapplication.cpp)

target_link_libraries(${CMAKE_PROJECT_NAME} android log)
```

Is the `import std`  feature supported? If yes, what am I doing wrong?

**Comment by DanAlbert** [comment, marked as answer]
> (Clang version is 19.0.0).

Not actually. It's an in-development snapshot of LLVM 19 (there's a file in the compiler directory called `AndroidVersion.txt` or something like that which explains the details). It may be older than 18.1.2. Without knowing which specific LLVM changes CMake needs (I'd be very shocked if anyone knows, that'd be a lot of unnecessary work on their part), the answer is probably "wait for NDK r29". If it doesn't work in r29, file a bug with a repro case. Not really sure if this is a bug or a feature request, but we can start it as a bug and recategorize if needed.

**Comment by youngjackly** [comment]
yes, you can import std with android. here is my way.
1. copy the ubuntu system's libc++ provided std.cppm and std.compat.cppm, with those two file and files the two file included to a directory. 
2. add a CMakelists.txt in the directory, build those files as a static lib.
3. then your app denpendend the static lib.
4. some function can't be build, comment it out. then in most situation, it can work.   the least situation, just include the header.
 hope help you.  My english is  spell is poor.  I just can read.  I know some word's spell is wrong, hope you can understand.

**Comment by youngjackly** [comment]
yes, you can import std with android. here is my way.
1. copy the ubuntu system's libc++ provided std.cppm and std.compat.cppm, with those two file and files the two file included to a directory. 
2. add a CMakelists.txt in the directory, build those files as a static lib.
3. then your app denpendend the static lib.
4. some function can't be build, comment it out. then in most situation, it can work.   the least situation, just include the header.
 hope help you.  My english is  spell is poor.  I just can read.  I know some word's spell is wrong, hope you can understand.

**Reply by LinHuan86** [reply]
could you give an example?

**Reply by youngjackly** [reply]
there is the CMakeLists.txt:
add_library(cxxmodule STATIC)

target_sources(cxxmodule PUBLIC FILE_SET CXX_MODULES FILES
	std.cppm
)
target_compile_options(cxxmodule PRIVATE -Wno-reserved-identifier -Wno-reserved-module-identifier)

don't build the std.compat.cppm, it's two much error,  this file have much lessor error, just about 10.  comment out where the error occoured.

in you target which used import std,  add this: target_link_libraries(xxxx PRIVATE  cxxmodule)
